modules/miner/blockmanager.go

```diff
@@ -117,13 +149,28 @@ func (m *Miner) SubmitHeader(bh types.BlockHeader) error {
 	lookupBH := bh
 	lookupBH.Nonce = zeroNonce
 	lockID := m.mu.Lock()
-	b, exists := m.blockMem[lookupBH]
-	m.mu.Unlock(lockID)
-	if !exists {
+	b, bExists := m.blockMem[lookupBH]
+	randTxn, txnExists := m.randTxnMem[lookupBH]
+
+	if !bExists || !txnExists {
+		m.mu.Unlock(lockID)
 		err := errors.New("block header returned late - block was cleared from memory")
 		m.log.Println("ERROR:", err)
 		return err
 	}
-	b.Nonce = bh.Nonce
-	return m.SubmitBlock(b)
+	// Reset block memory
+	m.memProgress = 0
```

Comment:
Had to think about this a bit.

I think that you should not delete the old memory. If a miner calls submitHeader for a block that is old (which means that an orphan was mined), a special 'stale err' should be returned so that the miner knows it's happening. I also don't think you need to set the memProgress to 0, you can just keep going from where you currently are. You just have to make sure that you get a new blockForWork at the current memory location, which is not necessarily divisible by anything.

Setting it to '0' is a convenient way to do that, but it might be harmful if it means miners are going to have less warning that they are occasionally mining stale blocks.